<img src="https://www.uvic.cat/sites/default/files/logo_3linies_uvic_color.jpg" alt="UVic Logo" width="200"/> <img src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSKMKJOIj0vlNde2xe47xQ-u_BLfb1xJtnCbg&s" alt="VHIO Logo" width="208" align="right"/>


<img alt="Python" src="https://img.shields.io/badge/-Python-3776AB?style=flat-square&logo=python&logoColor=white" /> <img alt="R" src="https://img.shields.io/badge/-R-276DC3?style=flat-square&logo=r&logoColor=white" /> <img alt="Jupyter Notebook" src="https://img.shields.io/badge/-JupyterNotebook-F37626?style=flat-square&logo=jupyter&logoColor=black" />

# SINGLE CELL COMPUTATIONAL TOOLS COMPARISON TO DETECT EPITHELIAL TUMORAL CELLS 

## Repository of single-cell analysis 

Single-cell RNA sequencing offers a unique view of the transcriptome. With a lot of different applications on this project, we have the objective of identifying the different cell types and tumoral cells on all the datasets with the Python tool to process single-cell RNA sequencing data, Scanpy. The VHIOâ€™s bioinformatics team has performed the same analysis, but with Seurat, the equivalent tool from R, another programming language. Another objective of this project is to compare the two more commonly used packages to analyze this data with both pipelines in order to find the most efficient way to work with single-cell data.
 



## Images used:

Python image -> martipuchal/scanpy_image

R image -> lgallagh/seurat
